calblueprint/wecaresolar
# Set up Firebase, Firebase Hosting, and push notifications

## 🖋 PR Description

Sets up Firebase for use in our app, along with Firebase Hosting (for deployment) and Cloud Messaging (for push notifications).

### Firebase Hosting
* Firebase hosting provides free hosting with HTTPS for our project. Our project is currently available at this link: https://we-care-solar-c3e5f.web.app/ (we can customize this later). Running `npm run build && firebase deploy` will deploy your current branch to this public URL _(don't do this though! we should deploy from master only)_
* Once this PR is merged in, **master will be automatically deployed to the main URL** whenever any future changes get merged. Also, **all open PRs will be automatically deployed to their own staging URLs** for easy testing! Staging URLs will be available for 30 days.
* As an example, you can test out this branch on any device at https://we-care-solar-c3e5f--push-notifications-wiwoxji6.web.app/

### Push Notifications
* Set up push notifications through Cloud Messaging. We can now ask the user for permission to send notifications,
request/save the push token associated with each device, and send notifications using the [Firebase Notifications UI](https://console.firebase.google.com/u/1/project/we-care-solar-c3e5f/notification)!
* **NOTE:** **Firebase push notifications only work over HTTPS** and localhost. This means we need to make sure to only deploy our app through Firebase Hosting (which should be done automatically now)

### 🤳Screenshots (if applicable)
Example on Android simulator of approving push notifications, exiting the app, then receiving a notification:
https://www.loom.com/share/a649211a4d6a482f891486d52bb91af0

### 🐞 Bug Report

* **The `firebase-messaging-sw.js` service worker file had to be placed in the `public/` folder in order to be registered properly as a service worker.** Otherwise, it was making a request to `localhost:3000/firebase-messaging-sw.js`, which by default would redirect us to the `App` component and go through React Router instead. Might need to keep this in mind for future service worker stuff!
* Push notifications require HTTPS, so we can only run this app either on localhost or a Firebase Hosting URL. This was the reason for setting up Firebase Hosting auto-deploys in this PR.

### 🧪 Testing

**Push notifications: Testing on a laptop**
* Open [this link](https://we-care-solar-c3e5f--push-notifications-wiwoxji6.web.app/) in one browser window. _Minimize this window so that it's considered in the "background"_
* Open another browser window and go to the [Notifications UI](https://console.firebase.google.com/u/1/project/we-care-solar-c3e5f/notification) in the Firebase console
* Click "Send your first message", enter a title and text, then go through the steps on the screen to send
* You should see a browser notification with the message! Clicking on it will take you to the app in your other window
* If you instead send the push while the app window is open, you can see the message contents in the Console (no notification will appear)

**Push notifications: Testing on Android simulator**
* Open [this link](https://we-care-solar-c3e5f--push-notifications-wiwoxji6.web.app/) in Chrome inside the simulator, then leave the app
* On your laptop, go to the [Notifications UI](https://console.firebase.google.com/u/1/project/we-care-solar-c3e5f/notification) in the Firebase console
* Click "Send your first message", enter a title and text, then go through the steps on the screen to send
* You should hear a sound on the Android simulator!
* Inside the simulator, pull down from the top of the screen to open the notification center. You should see the notification here! Clicking on it will take you to the app

**Firebase Hosting deploys**
* Check the example comment on this PR made by "github-actions" for the auto generated staging URL. Once this PR is merged, we'll get the same behavior on all of our branches with open PRs
* Once merged, make sure to verify that it deploys master to https://we-care-solar-c3e5f.web.app/

cc: @joelenelatief
